mindmeld/components/entity_resolver.py

```diff
@@ -562,38 +728,461 @@ def _predict_exact_match(self, entity):
+            num_layers = model.n_layers
+        elif hasattr(model, "layer"):  # eg. xlnet
+            num_layers = len(model.layer)
+        elif hasattr(model, "encoder"):  # eg. bert
+            num_layers = len(model.encoder.layer)
+        elif hasattr(model, "transformer"):  # eg. sentence_transformers models
+            num_layers = len(model.transformer.layer)
+        else:
+            raise ValueError(f"Not supported model {model} to obtain number of layers")
+
+        return num_layers
+
+    @staticmethod
+    def _text_length(text):
+        """
+        Help function to get the length for the input text. Text can be either
+        a list of ints (which means a single text as input), or a tuple of list of ints
+        (representing several text inputs to the model).
+
+        Union[List[int], List[List[int]]]
+        """
+        if isinstance(text, dict):
+            return len(next(iter(text.values())))
+        elif len(text) == 0 or isinstance(text[0], int):
+            return len(text)
+        else:
+            return sum([len(t) for t in text])
+
+    @staticmethod
+    def _batch_to_device(batch, target_device):
+        """
+        send a pytorch batch to a device (CPU/GPU)
+        """
+        tensor = _get_module_or_attr("torch", "Tensor")
+        for key in batch:
+            if isinstance(batch[key], tensor):
+                batch[key] = batch[key].to(target_device)
+        return batch
+
+    @staticmethod
+    def _compute_cosine_similarity(syn_embs, entity_emb, return_as_dict=False):
+        """Uses cosine similarity metric on synonym embeddings to sort most relevant ones
+            for entity resolution
+
+        Args:
+            syn_embs (dict): a dict of synonym and its corresponding embedding from bert
+            entity_emb (np.array): embedding of the input entity text, an array of size 1
+        Returns:
+            Union[dict, list[tuple]]: if return_as_dict, returns a dictionary of synonyms and their
+                                        scores, else a list of sorted synonym names, paired with
+                                        their similarity scores (descending)
+        """
+
+        entity_emb = entity_emb.reshape(1, -1)
+        synonyms, synonyms_encodings = zip(*syn_embs.items())
+        similarity_scores = cosine_similarity(np.array(synonyms_encodings), entity_emb).reshape(-1)
+        similarity_scores = np.around(similarity_scores, decimals=2)
+
+        if return_as_dict:
+            return dict(zip(synonyms, similarity_scores))
+
+        # results in descending scores
+        return sorted(list(zip(synonyms, similarity_scores)), key=lambda x: x[1], reverse=True)
+
+    def _fit(self, clean):
+        """
+        Fits the resolver model
+
+        Args:
+            clean (bool): If ``True``, deletes existing dump of synonym embeddings file
+        """
+
+        cache_path = self.cache_path(hash_json=self.er_config)
```

Comment:
it's great to have support on fitting model incrementally! One question here: Since we also consider config settings when generating hash for the cache file we could have multiple cache files for different set of config settings. I believe we would only remove the cache file with the current set when `clean=True` right?

Reply:
Thats correct. Only when explicitly asked for, we delete the cache file.

Reply:
sorry I realized my comment wasn't clear. I meant there could be multiple cache files (based on different config) and only the one matching the current config will be removed right?

Reply:
Yes, that is correct. Only the one with current config will be deleted.